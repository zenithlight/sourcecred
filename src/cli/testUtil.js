// @flow

import {type Command, type ExitCode, handlingErrors} from "./command";

// Run a command, capturing its stdout, stderr, and exit code. A thrown
// exception will be handled as with `handlingErrors`.
export async function run(
  command: Command,
  args: $ReadOnlyArray<string>
): Promise<{|
  +exitCode: ExitCode,
  +stdout: $ReadOnlyArray<string>,
  +stderr: $ReadOnlyArray<string>,
|}> {
  const stdout = [];
  const stderr = [];
  const exitCode = await handlingErrors(command)(args, {
    out: (x) => void stdout.push(x),
    err: (x) => void stderr.push(x),
  });
  return {exitCode, stdout, stderr};
}
