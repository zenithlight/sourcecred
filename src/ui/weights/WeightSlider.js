// @flow

import React, {type Node as ReactNode} from "react";
import {Grid, Slider, Tooltip} from "@material-ui/core";

/**
 * The Weight that the user supplies via a WeightSlider.
 * Weights are finite and non-negative. The vast majority of weights cannot
 * be selected via the WeightSlider, which is only capable of choosing a few
 * specific weights.
 */
export type Weight = number;

/**
 * SliderPosition is an integer in the range `[MIN_SLIDER, MAX_SLIDER]` which
 * describes the current position of the WeightSlider's slider.
 */
export type SliderPosition = number;

export const MIN_SLIDER: SliderPosition = -5;
export const MAX_SLIDER: SliderPosition = 5;

export type Props = {|
  // The current Weight the WeightSlider is set to
  // Note: It is acceptable to pass the WeightSlider a weight which the slider
  // is incapable of picking (e.g. 1/3).
  +weight: Weight,
  // The name associated with the slider; displayed adjacent to the slider.
  +name: React$Node,
  // Callback invoked with the new weight after the user adjusts the slider.
  +onChange: (Weight) => void,
  +description: string,
|};

/**
 * The WeightSlider is a [controlled component] which lets the user choose a
 * numeric weight by dragging a slider.
 *
 * The weight varies exponentially in the slider position, so that moving the
 * slider by one notch changes the weight by a power of two. The exception is
 * when the user drags the slider to its minimum value, in which case the
 * weight is set to 0.
 *
 * In addition to rendering a slider (instantiated as a [range input]), it also renders
 * a description of the slider, and the weight associated with the current slider position.
 *
 * Note that it is possible to configure the WeightSlider with a weight that
 * that is impossible to select via the slider; for example, 1/3. In such cases,
 * the WeightSlider's slider will be set to the closest possible position, and after
 * the user adjusts the weight via the slider, it will always correspond exactly to
 * the slider position.
 *
 * [controlled component]: https://reactjs.org/docs/forms.html#controlled-components
 * [range input]: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/range
 */
export class WeightSlider extends React.Component<Props> {
  render(): ReactNode {
    return (
      <Tooltip title={this.props.description} placement="top">
        <Grid container justify="space-between">
          <Grid item xs={4}>
            {this.props.name}
          </Grid>
          <Grid item xs={5}>
            <Slider
              value={weightToSlider(this.props.weight)}
              min={MIN_SLIDER}
              max={MAX_SLIDER}
              step={1}
              valueLabelDisplay="off"
              onChange={(_, val) => {
                const weight: Weight = sliderToWeight(val);
                this.props.onChange(weight);
              }}
            />
          </Grid>
          <Grid item xs={1}>
            {formatWeight(this.props.weight)}
          </Grid>
        </Grid>
      </Tooltip>
    );
  }
}

/**
 * Converts a Weight into a SliderPosition.
 *
 * If the Weight is an illegal value (negative, NaN, or infinite), an error
 * will be thrown.
 *
 * It is possible that the Weight will be a legal value, but will not
 * correspond to any SliderPosition (as SliderPositions are always integers, in
 * the range `[MIN_SLIDER, MAX_SLIDER]`). In such cases, it may seem principled
 * to throw a conversion error. However, initial weights are provided by
 * plugins, and plugin authors might reasonably choose weights like 1/3 that do
 * not correspond to a slider value. We choose to make the WeightSlider robust
 * to such decisions by rounding to the closest appropriate slider position. In
 * such cases, the WeightSlider's weight will be inconsistent with the slider
 * position, and moving the Slider away from and back to its initial position
 * will actually change the weight. This is unfortunate, but is preferable to
 * having the application error.
 */
export function weightToSlider(weight: Weight): SliderPosition {
  if (!isFinite(weight) || weight < 0) {
    throw new Error(`illegal weight: ${weight}`);
  }
  const tentativePosition = Math.log2(weight);
  if (tentativePosition < MIN_SLIDER) {
    return MIN_SLIDER;
  }
  if (tentativePosition > MAX_SLIDER) {
    return MAX_SLIDER;
  }
  return Math.round(tentativePosition);
}

/**
 * Converts a SliderPosition into a Weight.
 *
 * SliderPosition must be an integer in the range `[MIN_SLIDER, MAX_SLIDER]`.
 * If it is not, then an error will be thrown.
 *
 * Over the domain of legal slider positions `p`, `weightToSlider(sliderToWeight(p))`
 * is the identity function.
 */
export function sliderToWeight(sliderPosition: SliderPosition): Weight {
  if (sliderPosition < MIN_SLIDER || sliderPosition > MAX_SLIDER) {
    throw new Error(`Slider position out of range: ${sliderPosition}`);
  }
  if (isNaN(sliderPosition)) {
    throw new Error("illegal value: NaN");
  }
  if (Math.round(sliderPosition) !== sliderPosition) {
    throw new Error(`slider position not integer: ${sliderPosition}`);
  }
  return sliderPosition === MIN_SLIDER ? 0 : 2 ** sliderPosition;
}

export function formatWeight(n: number): string {
  if (n < 0 || !isFinite(n)) {
    throw new Error(`Invalid weight: ${n}`);
  }
  if (n >= 1 || n === 0) {
    return n.toFixed(0) + "×";
  } else {
    return `1/${(1 / n).toFixed(0)}×`;
  }
}
