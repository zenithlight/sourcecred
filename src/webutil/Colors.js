// @flow

import deepFreeze from "deep-freeze";
export type HexColor = string;

export default (deepFreeze({
  brand: {
    medium: "#0872A2",
    dark: "#3A066A",
  },
  accent: {
    medium: "#FF3201",
  },
}): {|
  +brand: {|
    +medium: HexColor,
    +dark: HexColor,
  |},
  +accent: {|
    +medium: HexColor,
  |},
|});
