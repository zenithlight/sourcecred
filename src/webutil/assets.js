// @flow

import normalize from "../util/pathNormalize";

/**
 * Resolver for static assets (e.g., images, PDFs) and API data (e.g.,
 * the repository registry, plugin data). Any references to resources
 * should be resolved through this API.
 */
export class Assets {
  +_root: ?string;

  /**
   * Construct a resolver given a path to the root of the site. This can
   * be a relative path, like `../..`, or an absolute path, like `/`.
   */
  constructor(root: ?string) {
    this._root = root == null ? root : normalize(root);
  }

  _getRoot(): string {
    if (this._root == null) {
      throw new Error("asset root path uninitialized");
    }
    return this._root;
  }

  /**
   * Resolve a path as if the current directory and "/" both represent
   * the site root. For instance, "foo", "/foo", and "./foo" all
   * represent the same file. It is an error to specify a file that is
   * above the root, like "../bad".
   */
  resolve(path: string): any {
    if (normalize(path.replace(/^\/+/, "")).startsWith("..")) {
      // It doesn't make sense to traverse past the site's root. This is
      // likely an error in the caller.
      throw new Error("path outside site root: " + path);
    }
    return normalize(`${this._getRoot()}/${path}`);
  }
}

/**
 * Given an absolute path `p`, return a relative path `r` such that a
 * web page at pathname `p` should use `r` to refer to the root of the
 * application. The result will only contain components `.` and `..`.
 *
 * Examples:
 *
 *   - "/foo/" maps to "..";
 *   - "/foo/bar" also maps to "..";
 *   - "/foo/bar/" maps to "../..";
 *   - "/" maps to ".".
 *
 * If the argument does not start with "/", an error will be thrown.
 */
export function rootFromPath(path: string): any {
  const normalized = normalize(path);
  if (normalized[0] !== "/") {
    throw new Error("expected absolute path: " + JSON.stringify(path));
  }
  const levels = (normalized.match(/\//g) || []).length;
  return normalize(new Array(levels - 1).fill("..").join("/"));
}
