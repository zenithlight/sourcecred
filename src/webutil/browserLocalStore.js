// @flow

import {LocalStore} from "./localStore";

/*
 * A simple abstraction over 'localStorage' to provide transparent JSON
 * serialization and deserialization.
 *
 * The implementation is borrowed heavily from Khan Academy's LocalStore
 * module, and also KaVideoPlayer's SafeLocalStore module.
 */
export default class BrowserLocalStore implements LocalStore {
  version: string;
  keyPrefix: string;

  constructor({version, keyPrefix}: {|+version: string, +keyPrefix: string|}) {
    this.version = version;
    this.keyPrefix = keyPrefix;
  }

  cacheKey(key: string): string {
    if (!key) {
      throw new Error("Falsy key provided to cacheKey: " + key);
    }
    return [this.keyPrefix, this.version, key].join(":");
  }

  get<T>(key: string, whenUnavailable: T): T {
    if (!this.isEnabled()) {
      return whenUnavailable;
    }
    try {
      const data = window.localStorage[this.cacheKey(key)];
      if (data) {
        return JSON.parse(data);
      } else {
        return whenUnavailable;
      }
    } catch (e) {
      // If we had trouble retrieving, like FF's NS_FILE_CORRUPTED:
      // http://stackoverflow.com/q/18877643/
      return whenUnavailable;
    }
  }

  set(key: string, data: mixed): void {
    if (!this.isEnabled()) {
      return;
    }
    const stringified = JSON.stringify(data);

    window.localStorage[this.cacheKey(key)] = stringified;
  }

  /*
   * Delete whatever data was associated with the given key.
   */
  del(key: string): void {
    if (!this.isEnabled()) {
      return;
    }
    const cacheKey = this.cacheKey(key);
    if (cacheKey in window.localStorage) {
      // (IE throws when deleting a non-existent entry.)
      delete window.localStorage[cacheKey];
    }
  }

  /*
   * Local storage might be disabled in old browsers or in Safari's
   * private browsing mode. Don't die.
   */
  isEnabled(): boolean {
    const uid = String(+new Date());
    try {
      window.sessionStorage[uid] = uid;
      const enabled = window.sessionStorage[uid] === uid;
      window.sessionStorage.removeItem(uid);
      return enabled;
    } catch (e) {
      return false;
    }
  }
}
