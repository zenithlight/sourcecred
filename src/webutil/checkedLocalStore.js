// @flow

import deepEqual from "lodash.isequal";

import {LocalStore} from "./localStore";

/*
 * A wrapper for a `LocalStore` that adds contract checking. This
 * implementation verifies that all provided keys are strings and all
 * provided values are plain old JSON values.
 */
export default class CheckedLocalStore implements LocalStore {
  _base: LocalStore;

  constructor(base: LocalStore): void {
    this._base = base;
  }

  get<T>(key: string, whenUnavailable: T): T {
    this._validateKey(key);
    return this._base.get(key, whenUnavailable);
  }

  set(key: string, data: mixed): void {
    this._validateKey(key);
    this._validateValue(data);
    return this._base.set(key, data);
  }

  del(key: string): void {
    this._validateKey(key);
    return this._base.del(key);
  }

  _validateKey(key: string) {
    if (typeof key !== "string") {
      throw new Error(`bad key (${typeof key}): ${key}`);
    }
  }

  _validateValue(data: any) {
    try {
      if (deepEqual(data, JSON.parse(JSON.stringify(data)))) {
        return;
      }
    } catch (_) {} // eslint-disable-line no-empty
    throw new Error(`bad value: ${data}`);
  }
}
