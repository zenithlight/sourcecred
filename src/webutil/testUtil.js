// @flow

export function configureEnzyme() {
  const Enzyme = require("enzyme");
  const Adapter = require("enzyme-adapter-react-16");
  Enzyme.configure({adapter: new Adapter()});
  beforeEach(() => {
    // $FlowExpectedError[cannot-write]
    console.error = jest.fn();
    // $FlowExpectedError[cannot-write]
    console.warn = jest.fn();
  });
  afterEach(() => {
    expect(console.warn).not.toHaveBeenCalled();
    expect(console.error).not.toHaveBeenCalled();
  });
}

export function configureAphrodite() {
  const {StyleSheetTestUtils} = require("aphrodite/no-important");
  beforeEach(() => {
    StyleSheetTestUtils.suppressStyleInjection();
  });

  afterEach(() => {
    StyleSheetTestUtils.clearBufferAndResumeStyleInjection();
  });
}
