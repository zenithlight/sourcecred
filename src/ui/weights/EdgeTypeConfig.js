// @flow

import React, {type Node as ReactNode} from "react";
import {WeightSlider, type Props as WeightSliderProps} from "./WeightSlider";

import type {EdgeType} from "../../analysis/types";
import type {EdgeWeight} from "../../core/weights";

export class EdgeTypeConfig extends React.Component<{
  +weight: EdgeWeight,
  +type: EdgeType,
  +onChange: (EdgeWeight) => void,
}> {
  render(): ReactNode {
    const {weight, type} = this.props;
    const {forwards, backwards} = weight;
    const {forwardName, backwardName, description} = type;
    return (
      <div>
        <EdgeWeightSlider
          name={backwardName}
          weight={forwards}
          description={description}
          onChange={(newForwards) => {
            this.props.onChange({forwards: newForwards, backwards});
          }}
        />
        <EdgeWeightSlider
          name={forwardName}
          weight={backwards}
          description={description}
          onChange={(newBackwards) => {
            this.props.onChange({
              forwards,
              backwards: newBackwards,
            });
          }}
        />
      </div>
    );
  }
}

export function styledVariable(letter: string): ReactNode {
  return (
    // marginRight accounts for italicization
    <span style={{fontWeight: 700, fontStyle: "italic", marginRight: "0.15em"}}>
      {letter}
    </span>
  );
}

export class EdgeWeightSlider extends React.Component<WeightSliderProps> {
  render(): ReactNode {
    const modifiedName = (
      <React.Fragment>
        {styledVariable("α")} {this.props.name} {styledVariable("β")}
      </React.Fragment>
    );
    return (
      <WeightSlider
        name={modifiedName}
        weight={this.props.weight}
        description={this.props.description}
        onChange={this.props.onChange}
      />
    );
  }
}
