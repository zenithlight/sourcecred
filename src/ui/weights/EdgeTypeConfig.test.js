// @flow

import React from "react";
import {shallow} from "enzyme";

import {WeightSlider} from "./WeightSlider";
import {EdgeTypeConfig, EdgeWeightSlider} from "./EdgeTypeConfig";
import {assemblesEdgeType} from "../../plugins/demo/declaration";

require("../../webutil/testUtil").configureEnzyme();

describe("ui/weights/EdgeTypeConfig", () => {
  describe("EdgeTypeConfig", () => {
    function example() {
      const onChange = jest.fn();
      const type = assemblesEdgeType;
      const weight = {forwards: 1, backwards: 0.5};
      const element = shallow(
        <EdgeTypeConfig onChange={onChange} type={type} weight={weight} />
      );
      const forwardSlider = element.find(EdgeWeightSlider).at(0);
      const backwardSlider = element.find(EdgeWeightSlider).at(1);
      return {onChange, weight, type, forwardSlider, backwardSlider};
    }
    it("sets up the forward weight slider", () => {
      const {weight, type, forwardSlider} = example();
      expect(forwardSlider.props().name).toBe(type.backwardName);
      expect(forwardSlider.props().weight).toBe(weight.forwards);
    });
    it("sets up the backward weight slider", () => {
      const {weight, type, backwardSlider} = example();
      expect(backwardSlider.props().name).toBe(type.forwardName);
      expect(backwardSlider.props().weight).toBe(weight.backwards);
    });
    it("has a description", () => {
      const {backwardSlider} = example();
      expect(backwardSlider.props().description).toBe(
        assemblesEdgeType.description
      );
    });
    it("forward weight slider onChange works", () => {
      const {weight, forwardSlider, onChange} = example();
      forwardSlider.props().onChange(9);
      const updated = {backwards: weight.backwards, forwards: 9};
      expect(onChange).toHaveBeenCalledTimes(1);
      expect(onChange.mock.calls[0][0]).toEqual(updated);
    });
    it("backward weight slider onChange works", () => {
      const {weight, backwardSlider, onChange} = example();
      backwardSlider.props().onChange(9);
      const updated = {forwards: weight.forwards, backwards: 9};
      expect(onChange).toHaveBeenCalledTimes(1);
      expect(onChange.mock.calls[0][0]).toEqual(updated);
    });
  });
  describe("EdgeWeightSlider", () => {
    function example() {
      const onChange = jest.fn();
      const element = shallow(
        <EdgeWeightSlider
          weight={3}
          name="foo"
          description="Description for test slider"
          onChange={onChange}
        />
      );
      const weightSlider = element.find(WeightSlider);
      return {element, onChange, weightSlider};
    }
    it("renders the name along with some Greek characters", () => {
      const {weightSlider} = example();
      const name = weightSlider.props().name;
      expect(name).toMatchSnapshot();
    });
    it("passes through the weight unchanged", () => {
      const {weightSlider} = example();
      expect(weightSlider.props().weight).toBe(3);
    });
    it("onChange is wired properly", () => {
      const {weightSlider, onChange} = example();
      expect(weightSlider.props().onChange).toBe(onChange);
    });
  });
});
