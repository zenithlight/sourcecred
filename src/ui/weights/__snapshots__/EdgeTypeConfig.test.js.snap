// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ui/weights/EdgeTypeConfig EdgeWeightSlider renders the name along with some Greek characters 1`] = `
<React.Fragment>
  <span
    style={
      Object {
        "fontStyle": "italic",
        "fontWeight": 700,
        "marginRight": "0.15em",
      }
    }
  >
    α
  </span>
   
  foo
   
  <span
    style={
      Object {
        "fontStyle": "italic",
        "fontWeight": 700,
        "marginRight": "0.15em",
      }
    }
  >
    β
  </span>
</React.Fragment>
`;
